glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,164 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! # Metrics Ping Scheduler
+//!
+//! The Metrics Ping Scheduler (MPS) is responsible for scheduling "metrics" pings.
+//! It implements the spec described in
+//! [the docs](https://mozilla.github.io/glean/book/user/pings/metrics.html#scheduling)
+
+use crate::metrics::{DatetimeMetric, StringMetric, TimeUnit};
+use crate::{CommonMetricData, Glean, Lifetime, INTERNAL_STORAGE};
+use chrono::prelude::*;
+use once_cell::sync::OnceCell;
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+
+const SCHEDULED_HOUR: u32 = 4;
+
+static THREAD_GENERATION: OnceCell<Arc<AtomicU32>> = OnceCell::new();
+
+/// Performs startup checks to decide when to schedule the next "metrics" ping collection.
+/// Must be called before draining the preinit queue.
+/// (We're at the Language Bindings' mercy for that)
+pub fn schedule(glean: &Glean) {
+    let now = Local::now();
+
+    schedule_internal(&glean, now)
+}
+
+fn schedule_internal(glean: &Glean, now: DateTime<Local>) {
+    let last_sent_build_metric = get_last_sent_build_metric();
+    if let Some(last_sent_build) = last_sent_build_metric.get_value(&glean, INTERNAL_STORAGE) {
+        // XXX: If `app_build` is longer than StringMetric's max length, we will always
+        // treat it as a changed build when really it isn't.
+        if &last_sent_build != glean.app_build.as_ref().unwrap() {
+            last_sent_build_metric.set(&glean, glean.app_build.as_ref().unwrap());
+            log::info!("Builds don't match. Sending 'metrics' ping");
+            collect_ping_and_reschedule(&glean, now, Some("upgrade"));
+            return;
+        }
+    }
+
+    let last_sent_time = get_last_sent_time_metric().get_value(&glean, INTERNAL_STORAGE);
+    if let Some(last_sent) = last_sent_time {
+        log::info!("The 'metrics' ping was last sent on {}", last_sent);
+    }
+
+    // We aim to cover 3 cases here:
+    //
+    // 1. The ping was already collected on the current calendar day;
+    //    only schedule one for collection on the next calendar day at the due time.
+    // 2. The ping was NOT collected on the current calendar day AND we're later
+    //    than today's due time; collect the ping immediately.
+    // 3. The ping was NOT collected on the current calendar day BUT we still have
+    //    some time to the due time; schedule for sending the current calendar day.
+
+    let already_sent_today = last_sent_time.map_or(false, |d| d.ordinal() == now.ordinal());
```

Comment:
`ordinal` gives day of year: https://docs.rs/chrono/0.4.19/chrono/trait.Datelike.html#tymethod.ordinal

Reply:
Guess I looked at ... I don't know what I looked at.
Anyway, bug stands for a browser that's not used in a full year! (Edge case I guess...)

Reply:
How 'bout I put in a comment about it then? I trust `chrono` to properly handle getting me calendar days since Jan 1 over me trying to figure out days since an epoch. I'd get the timezones wrong for sure.

Reply:
👍